File: matlab/path_automatron.m
function [path_para,temp_act_1,temp_act_2]=path_automatron(path_para,node_act_1,node_act_2)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% This function update the status of a single path
%
% Inputs:
% path_para: Cell array, parameters for the paths
%
%            format: {'path_name',path_state_index, entry_node_index,
%            exit_node_index, amplitude_factor, forward_speed,
%            backward_speed, forward_timer_current, forward_timer_default,
%            backward_timer_current, backward_timer_default, path_length,
%            path_slope}
% node_act_1: boolean, activation status of the entry node
% node_act_2: boolean, activation status of the exit node
%
% Outputs:
% temp_act_1: boolean, local temporary node activation of the entry node
% temp_act_2: boolean, local temporary node activation of the exit node
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

temp_act_1=0;
temp_act_2=0;
switch path_para{2}
    case 1 % Idle
        % if activation coming from entry node
        if node_act_1
            % Antegrade conduction
            path_para{2}=2;
        % if activation coming from exit node
        else if node_act_2
                % Retrograde conduction
                path_para{2}=3;
            end
        end
    case 2 % Antegrade conduction
        % if activation coming from exit node
        if node_act_2
            % double
            path_para{2}=4;
%             % reset timer
%             path_para{8}=path_para{9};
        else
            % if timer running out
            if path_para{8}==0
                % reset timer
                path_para{8}=path_para{9};
                % activate exit node
                temp_act_2=1;
                % go to conflict state
                path_para{2}=4;
            else
                % timer
                path_para{8}=path_para{8}-1;
            end
        end
            
    case 3 % Retro
        % if activation coming from entry node
        if node_act_1
            % conflict
            path_para{2}=4;
%             % reset timer
%             path_para{10}=path_para{11};
        else
            % if timer runs out
            if path_para{10}==0
                % reset timer
                path_para{10}=path_para{11};
                % activate the entry node
                temp_act_1=1;
                % change state to conflict
                path_para{2}=4;
            else
                % timer
                path_para{10}=path_para{10}-1;
            end
        end
    case 4 % Conflict
        % go to Idle state
        path_para{2}=5;
    case 5
        path_para{2}=1;
%     case 5 % double
%                
%         if path_para{10}==0
%                 % reset timer
%                 path_para{10}=path_para{11};
%                 % activate the entry node
%                 temp_act_1=1;
%                 % change state to conflict
%                 path_para{2}=2;
%                 return
%         end
%         if path_para{8}==0
%                 % reset timer
%                 path_para{8}=path_para{9};
%                 % activate exit node
%                 temp_act_2=1;
%                 % go to conflict state
%                 path_para{2}=3;
%                 return
%         end
%         if abs(1-path_para{8}/path_para{9}-path_para{10}/path_para{11})<0.9/min([path_para{9},path_para{11}])
%                 path_para{10}=path_para{11};
%                 path_para{8}=path_para{9};
%                 path_para{2}=4;
%                 
%         else
%             
%             path_para{8}=path_para{8}-1;
%             path_para{10}=path_para{10}-1;
%         end
        
end

File: frdm/path_automatron.cpp
/*This function is used update the states of the path based on the condition in the nodes
*/

void path_automatron(int *path_para,char node_act_1,char node_act_2,char tempActArray[])
{
    char temp_act_1 = 0;
    char temp_act_2 = 0;
    char state  = path_para[0];
    switch(state)
    {
        case 1:
        { //idle
            //if activation comng from entry node
            if(node_act_1)
            {
                //antegrade conduction
                path_para[0] = 2;
            }
            //activation from exit node
            else if(node_act_2)
             {
                //retrograde conduction
                path_para[0] = 3;
            }
            break;
        }
        case 2: { //antegrade conduction
            //if activation from exit node
            if(node_act_2)
             {
                //double
                path_para[0] = 4;
            }
            else
            {
                //if timer running out
                if(path_para[3] == 0)
                {
                    //reset timer
                    path_para[3] = path_para[4];
                    //activate exit node
                    temp_act_2 = 1;
                    //go to conflict state
                    path_para[0] = 4;
                }
                else
                 {
                    //timer
                    path_para[3] = path_para[3] - 1;
                }
            }
            break;
        }
        case 3: //retro
            // if activation coming from entry node
        {
            if(node_act_1)
             {
                path_para[0] = 4; //conflict
            }
             else
            {
                //if timer runs out
                if(path_para[5] == 0)
                 {
                    //reset timer
                    path_para[5] = path_para[6];
                    //activate entry node
                    temp_act_1 = 1;
                    //change state to conflict
                    path_para[0] = 4;
                }
                else
                 {
                    //timer
                    path_para[5] = path_para[5] - 1;
                }
            }
            break;
        }
        case 4: //conflict
        {
            //use state 5 to delay 2 ms
            path_para[0] = 5;
            break;
        }
        case 5:
        {
            path_para[0] = 1;
            break;
        }
    }
    tempActArray[0] = temp_act_1;
    tempActArray[1] = temp_act_2;
}
